### PR TITLE
Fix RNNs

### PR DESCRIPTION
**Pull request by rcurtin**
In #2777 I made some grevious errors when working on the BPTT implementation.  I am currently trying to work up an example to train RNNs to predict DGA domains for documentation, so these problems surfaced.  I'm not sure what I misunderstood about BPTT back then, but over the past ~4/5 weeks I have revamped the RNN implementation inside of mlpack and brought back numerous tests that had been in the `not_adapted/` directory (including the Reber grammar tests, which somehow did not make it into mlpack 4 at all).

### Design changes

 * A `RecurrentLayer` has some new expectations:
   - It must implement `RecurrentSize()`, which is the number of elements that need to be stored for recurrent state.
   - `Forward()` now must set the recurrent state for the current time step, which can be accessed with `RecurrentState(CurrentStep())`; however, if `AtFinalStep()` is `true`, then there is no need to do this.
   - `Forward()` can use the previous time step's state as `RecurrentState(PreviousStep())` (but should not change it!).
   - `Backward()` and `Gradient()` can use the current time step's recurrent gradient, accessed via `RecurrentGradient(CurrentStep())`.
   - The "recurrent gradient" is the derivative of the output of the network with respect to the current time step's hidden state; it's just like the `delta` that gets passed into `Backward()`, but via a recurrent connection.
   - If `HasPreviousStep()` is true, then `Backward()` should set the previous time step's recurrent gradient, accessed via `RecurrentGradient(PreviousStep())`.
   - If `AtFinalStep()` is `true`, then `Backward()` and `Gradient()` should not use `RecurrentGradient(CurrentStep())`.
   - An individual layer does not need to worry about allocating the memory to handle recurrent state or gradients; that is all done automatically by the base layer and the `RNN` class.

 * These changes above are documented in the code, and I will make a clear note of them when I finally write Markdown documentation for neural networks.
 
 * The `RNN` implementation of truncated BPTT used to add the output error at *every* time step during a single BPTT step (so, e.g., at time step `t`, we would compute the loss, backpropagate it, then at time step `t - 1` we would *also* compute the loss and backpropagate it *in addition to* the recurrent gradient from time step `t`, and so on and so forth).  Note that BPTT is already done at every time step, so this is redundant; when I made this change, my test networks started to converge *much* faster and matched output from PyTorch's RNNs and roughly matched numbers from papers I saw.  This 'bug' has persisted for a long time; I believe even mlpack's original RNN implementation had this issue.

### New features

 * I added the `LinearRecurrent` layer.  This one implements the "typical" recurrent cell, where the output is `Wx + Uh + b` where `x` is the non-recurrent input, `h` is the output from the last time step, and `b` is the bias.  Note that the output does *not* go through a sigmoid before passing through the recurrent connection... I can change that if anyone thinks it is an important idea.  (In practice this layer is only useful for super simple experiments and learning anyway; they don't learn anything reasonable on real data.)

 * I revived and reimplemented the `LSTM` class.  I used [this paper](https://arxiv.org/pdf/1503.04069) which made me very happy because I did not have to derive all the calculus myself.  Their 'default' LSTM has peephole connections, so I added those---this is a change from our previous implementation, and in my basic testing (and theirs too) it appears to help learning somewhat.  However, it is computationally an extra thing, so whenever optimization happens to this layer, it might make sense to allow peephole connections to be disabled.

 * I wrote 2100 lines of tests for the LSTM because I was so paranoid I would get it wrong.  These tests essentially check the forward, backward, and gradient passes to make sure they give the right outputs when only a subset of the weights or inputs are nonzero.  It's fairly exhaustive (and probably exhausting to read).

### Unrelated or other changes

 * `MeanSquaredError` by default wasn't implementation the mean squared error---it was just implementing the squared error.  I fixed this.

### Final notes

 * I used a PyTorch implementation of a character-RNN to implement the Reber grammar test in PyTorch and compare the output.  That helped me track down the last bug or two.  One of the things that I noticed was that the mlpack test would do 10 or 20 epochs in the time PyTorch would do only 1.  It's not a formal benchmark; I did not optimize any of the PyTorch code and there may be stupid things I'm doing all over.  But... I also didn't go through the mlpack LSTM implementation and optimize it!  So runtime-wise, I'm pretty happy, and I think there is a lot more potential.  The old mlpack had `FastLSTM` which would compress a number of the matrix multiplications involved with the LSTM into one.

Ok, that's enough... I have more thoughts but this is a very long PR description already.

### CODE REVIEW

**Review by shrit — APPROVED** [review]

File: src/mlpack/tests/ann/layer/lstm.cpp [hunk 1]
```diff
@@ -0,0 +1,2133 @@
+/**
+ * @file lstm.cpp
+ * @author Marcus Edel
+ * @author Praveen Ch
```
Comment (zoq) [inline src/mlpack/tests/ann/layer/lstm.cpp:4]: I think you should add yourself.

File: src/mlpack/tests/ann/layer/lstm.cpp [hunk 2]
```diff
@@ -0,0 +1,2133 @@
...
+  ens::StandardSGD opt(0.1, 1, maxEpochs * trainData.n_slices, -100, false);
+
+  network1->Train(trainData, trainLabels, opt);
+  network1->Predict(trainData, predictions1);
+
+  RNN<> network2 = *network1;
+  delete network1;
+
+  // Deallocating all of network1's memory, so that network2 does not use any
+  // of that memory.
```
Comment (zoq) [inline src/mlpack/tests/ann/layer/lstm.cpp:98]: Should this comment be above delete?

File: src/mlpack/tests/ann/layer/lstm.cpp [hunk 3]
```diff
@@ -0,0 +1,2133 @@
...
+  arma::cube predictions2;
+  ens::StandardSGD opt(0.1, 1, maxEpochs * trainData.n_slices, -100, false);
+
+  network1->Train(trainData, trainLabels, opt);
+  network1->Predict(trainData, predictions1);
+
+  RNN<> network2(std::move(*network1));
+  delete network1;
+
+  // Deallocating all of network1's memory, so that network2 does not use any
```
Comment (zoq) [inline src/mlpack/tests/ann/layer/lstm.cpp:120]: See comment above.

**Comment by eddelbuettel** [comment]
Any idea about the one failure?  Does the test need updating?

```
-------------------------------------------------------------------------------
PReLUIntegrationTest
-------------------------------------------------------------------------------
/home/runner/work/mlpack/mlpack/src/mlpack/tests/ann/layer/parametric_relu.cpp:104
...............................................................................

/home/runner/work/mlpack/mlpack/src/mlpack/tests/ann/layer/parametric_relu.cpp:144: FAILED:
  REQUIRE( success == true )
with expansion:
  false == true
```

**Comment by rcurtin** [comment]
I was unable to reproduce the failure locally---I tried 5000 different random seeds and none of them failed on this branch.  So at least to me it is a mystery, but, none of the code changes that I've made here should affect the `PReLU` layer.  I believe I have seen that error before on other builds and not been able to track down what is going on there either; so perhaps this is a mystery that will remain until another time?

File: src/mlpack/tests/ann/layer/lstm.cpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/tests/ann/layer/lstm.cpp:4]: Done in 6f3326730c.

File: src/mlpack/tests/ann/layer/lstm.cpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/tests/ann/layer/lstm.cpp]: Good point, done in 6f3326730c.

File: src/mlpack/tests/ann/layer/lstm.cpp (hunk 3 above)
Comment (rcurtin) [reply src/mlpack/tests/ann/layer/lstm.cpp]: Also done in 6f3326730c.

**Review by zoq — APPROVED** [review]
